File: packages/scraper/_protocol.py
#!/usr/bin/env python3
"""
Shared NDJSON protocol emitter for Python scraper and RPA scripts.
"""
from __future__ import annotations

import json
import sys
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

PROTOCOL_VERSION = "1.0"
MAX_MESSAGE_LENGTH = 2000


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def serialize_payload(payload: dict[str, Any]) -> str:
    """Serialize a protocol payload deterministically."""
    return json.dumps(payload, ensure_ascii=False, separators=(",", ":"), sort_keys=True)


@dataclass
class ProtocolEmitter:
    """Emit protocol-compliant events for RPA execution streams."""

    run_id: str
    sequence: int = field(default=0, init=False)

    def _event_base(self, event_type: str) -> dict[str, Any]:
        event = {
            "protocolVersion": PROTOCOL_VERSION,
            "runId": self.run_id,
            "sequence": self.sequence,
            "timestamp": _now_iso(),
            "eventType": event_type,
        }
        self.sequence += 1
        return event

    def emit_progress(self, event: dict[str, Any]) -> None:
        payload = self._event_base("progress")
        payload.update(event)
        sys.stderr.write(f"{serialize_payload(payload)}\n")
        sys.stderr.flush()

    def emit_result(self, event: dict[str, Any]) -> None:
        payload = self._event_base("result")
        payload["result"] = event
        sys.stdout.write(f"{serialize_payload(payload)}\n")
        sys.stdout.flush()

    def emit_error(self, code: str, message: str, details: dict[str, Any] | None = None) -> None:
        payload = self._event_base("error")
        safe_message = message.strip()[:MAX_MESSAGE_LENGTH] if message else "Unknown error"
        payload["error"] = {
            "code": code,
            "message": safe_message,
            "source": "python-script",
            **({"details": details} if details else {}),
        }
        sys.stdout.write(f"{serialize_payload(payload)}\n")
        sys.stdout.flush()

File: packages/scraper/src/runtime/protocol.ts
import type { ErrorEnvelope, RpaRunErrorCode } from "@bao/shared/schemas/error-envelope.schema";
import type { RpaRunEvent, RpaRunResult } from "@bao/shared/schemas/rpa-events.schema";
import { RPA_PROTOCOL_VERSION } from "@bao/shared/schemas/rpa-protocol.schema";
import type { JsonObject } from "@bao/shared/utils/json";

type ProtocolOutputStream = "stdout" | "stderr";

const writeProtocolLine = (stream: ProtocolOutputStream, event: RpaRunEvent): void => {
  const target = stream === "stdout" ? process.stdout : process.stderr;
  target.write(`${JSON.stringify(event)}\n`);
};

/**
 * Writes a single NDJSON protocol line to the requested stream.
 *
 * @param stream Output stream used for the event.
 * @param event Fully constructed protocol event.
 */
export const emitProtocolLine = (stream: ProtocolOutputStream, event: RpaRunEvent): void => {
  writeProtocolLine(stream, event);
};

/**
 * Contract-first protocol emitter for Bun-based job-apply automation scripts.
 */
export class ProtocolEmitter {
  private readonly runId: string;
  private sequence = 0;

  /**
   * Creates a protocol emitter for a single run identifier.
   *
   * @param runId Stable run identifier supplied by the server.
   */
  constructor(runId: string) {
    this.runId = runId;
  }

  private nextMetadata() {
    const metadata = {
      protocolVersion: RPA_PROTOCOL_VERSION,
      runId: this.runId,
      sequence: this.sequence,
      timestamp: new Date().toISOString(),
    } as const;
    this.sequence += 1;
    return metadata;
  }

  /**
   * Emits a progress event to stderr.
   *
   * @param event Progress payload without shared metadata.
   */
  emitProgress(
    event: Pick<
      Extract<RpaRunEvent, { eventType: "progress" }>,
      "action" | "status" | "message" | "step" | "totalSteps"
    >,
  ): void {
    emitProtocolLine("stderr", {
      ...this.nextMetadata(),
      eventType: "progress",
      ...event,
    });
  }

  /**
   * Emits a terminal result event to stdout.
   *
   * @param result Final automation result payload.
   */
  emitResult(result: RpaRunResult): void {
    emitProtocolLine("stdout", {
      ...this.nextMetadata(),
      eventType: "result",
      result,
    });
  }

  /**
   * Emits a terminal error event to stdout.
   *
   * @param code Runtime-neutral error classification.
   * @param message Human-readable error detail.
   * @param details Optional JSON details payload.
   */
  emitError(code: RpaRunErrorCode, message: string, details?: JsonObject): void {
    const error: ErrorEnvelope = {
      code,
      message,
      ...(details ? { details } : {}),
    };

    emitProtocolLine("stdout", {
      ...this.nextMetadata(),
      eventType: "error",
      error: {
        ...error,
        source: "@bao/scraper",
      },
    });
  }
}
